# Benvenuto/a nel Notebook per Imparare le Basi di Python!

Questo notebook ti guiderà attraverso i concetti fondamentali di Python in modo semplice e interattivo. Per ogni concetto, troverai una spiegazione chiara e un mini esercizio per mettere subito in pratica ciò che hai imparato.

## 1. Stampa: Il tuo primo programma Python!

Il comando `print()` è usato per visualizzare un messaggio o il valore di una variabile sullo schermo.

**Esercizio:** Scrivi una riga di codice che stampi il messaggio "Ciao, Python!"

In [1]:
# Scrivi qui il tuo codice
print("Ciao, Python!")

Ciao, Python!


**Riscontro atteso:** Dovresti vedere la scritta `Ciao, Python!` visualizzata qui sotto dopo aver eseguito la cella precedente (premendo Shift + Invio o cliccando sul pulsante "Esegui").

## 2. Variabili e Tipi di Dati

Le **variabili** sono come dei contenitori usati per memorizzare dati. Ogni variabile ha un nome e un valore.
Python ha diversi **tipi di dati** incorporati, i più comuni sono:
- **Stringhe (`str`):** Sequenze di caratteri, come testo. Si scrivono tra virgolette singole (`'`) o doppie (`"`). Esempio: `"ciao"`, `'mondo'`.
- **Numeri Interi (`int`):** Numeri interi, positivi o negativi, senza decimali. Esempio: `10`, `-3`.
- **Numeri con Virgola Mobile (`float`):** Numeri con la parte decimale. Esempio: `3.14`, `-0.5`.
- **Booleani (`bool`):** Rappresentano uno dei due valori: `True` (vero) o `False` (falso).

**Esercizio:**
1. Crea una variabile chiamata `saluto` e assegnale la stringa "Buongiorno".
2. Stampa il valore della variabile `saluto`.

In [2]:
# Scrivi qui il tuo codice
saluto = "Buongiorno"
print(saluto)

Buongiorno


**Riscontro atteso:** Dovresti vedere la scritta `Buongiorno`.

**Esercizio:**
1. Crea una variabile chiamata `eta` e assegnale un numero intero che rappresenta la tua età.
2. Stampa il valore della variabile `eta`.

In [3]:
# Scrivi qui il tuo codice
eta = 19
print(eta)

19


**Riscontro atteso:** Dovresti vedere il numero che hai inserito.

**Esercizio:**
1. Crea una variabile chiamata `prezzo` e assegnale un numero con la virgola, ad esempio `9.99`.
2. Stampa il valore della variabile `prezzo`.

In [4]:
# Scrivi qui il tuo codice
prezzo = 9.99
print(prezzo)

9.99


**Riscontro atteso:** Dovresti vedere il numero `9.99` (o quello che hai inserito).

**Esercizio:**
1. Crea una variabile chiamata `studente` e assegnale il valore booleano `True`.
2. Stampa il valore della variabile `studente`.

In [5]:
# Scrivi qui il tuo codice
studente = True
print(studente)

True


**Riscontro atteso:** Dovresti vedere `True`.

Puoi scoprire il tipo di una variabile usando la funzione `type()`.

**Esercizio:** Stampa il tipo della variabile `prezzo` creata precedentemente.

In [6]:
# Scrivi qui il tuo codice
type(prezzo)

float

**Riscontro atteso:** Dovresti vedere `<class 'float'>`.

### Extra: Casting di tipi di dato

**Il casting** permette di convertire una variabile da un tipo a un altro. Ad esempio, puoi trasformare un `int` in `float` o una `stringa` in un numero se compatibile.

Esempi comuni:
- Da intero a float: `float(5)`
- Da float a intero (scarta i decimali): `int(3.75)`
- Da stringa a intero: `int('23')` (solo se la stringa contiene cifre!)
- Da numero a stringa: `str(42)`


**Esercizio 1:** Dato un numero intero `n = 7`,
1. Converti `n` in float, assegnando il risultato alla variabile `n_float`.
2. Stampa valore e tipo di `n_float`.

In [7]:
# Scrivi qui il tuo codice
n = 7
n_float = float(n)
print(n_float)

7.0


**Riscontro atteso:**
7.0
<class 'float'>


**Esercizio 2:** Prova a convertire una stringa in int.
1. Crea la variabile `testo = '18'`.
2. Trasforma `testo` in numero intero e memorizza in `numero`.
3. Stampa tipo e valore di `numero`.

In [8]:
# Scrivi qui il tuo codice
testo = '18'
numero = int(testo)
print(numero)

18


**Riscontro atteso:**
<class 'int'>
18


## 3. Operatori

Gli operatori sono simboli speciali che eseguono operazioni su variabili e valori.

### Operatori Aritmetici
- `+` (addizione)
- `-` (sottrazione)
- `*` (moltiplicazione)
- `/` (divisione, restituisce un float)
- `//` (divisione intera, restituisce un int, scartando la parte decimale)
- `%` (modulo, restituisce il resto della divisione)
- `**` (esponenziazione)

**Esercizio:** Crea due variabili `a = 10` e `b = 3`. Stampa il risultato della loro somma.

In [9]:
# Scrivi qui il tuo codice
a = 10
b = 3
print(a+b)

13


**Riscontro atteso:** Dovresti vedere `13`.

**Esercizio:** Usando le stesse variabili `a` e `b`, stampa il risultato di `a` diviso `b` (divisione float).

In [10]:
# Scrivi qui il tuo codice 
print(a/b)

3.3333333333333335


**Riscontro atteso:** Dovresti vedere un numero simile a `3.3333333333333335`.

**Esercizio:** Usando le stesse variabili `a` e `b`, stampa il resto della divisione intera di `a` per `b`.

In [11]:
# Scrivi qui il tuo codice
print(a%b)

1


**Riscontro atteso:** Dovresti vedere `1`.

### Operatori di Confronto
Restituiscono un valore booleano (`True` o `False`).
- `==` (uguale a)
- `!=` (diverso da)
- `>` (maggiore di)
- `<` (minore di)
- `>=` (maggiore o uguale a)
- `<=` (minore o uguale a)

**Esercizio:** Usando le variabili `a = 10` e `b = 3`, stampa il risultato del confronto `a > b`.

In [12]:
# Scrivi qui il tuo codice 
print(a>b)

True


**Riscontro atteso:** Dovresti vedere `True`.

**Esercizio:** Stampa il risultato del confronto `a == b`.

In [13]:
# Scrivi qui il tuo codice
print(a==b)

False


**Riscontro atteso:** Dovresti vedere `False`.

### Operatori Logici
Usati per combinare espressioni condizionali.
- `and` (vero se entrambe le espressioni sono vere)
- `or` (vero se almeno una delle espressioni è vera)
- `not` (inverte il valore booleano)

**Esercizio:** Crea due variabili booleane `cond1 = True` e `cond2 = False`. Stampa il risultato di `cond1 and cond2`.

In [ ]:
# Scrivi qui il tuo codice

**Riscontro atteso:** Dovresti vedere `False`.

**Esercizio:** Usando `cond1` e `cond2`, stampa il risultato di `cond1 or cond2`.

In [ ]:
# Scrivi qui il tuo codice

**Riscontro atteso:** Dovresti vedere `True`.

## 4. Strutture di Controllo

Permettono di controllare il flusso di esecuzione del programma.

### `if`, `elif`, `else` (Condizionali)
Eseguono blocchi di codice diversi a seconda che una condizione sia vera o falsa.
L'indentazione (spazi all'inizio della riga) è FONDAMENTALE in Python per definire i blocchi di codice.

**Esercizio:** Crea una variabile `numero = 7`. Scrivi un blocco `if/else` che stampi "Il numero è positivo" se `numero` è maggiore di 0, altrimenti stampi "Il numero non è positivo".

In [ ]:
# Scrivi qui il tuo codice

**Riscontro atteso:** Dovresti vedere `Il numero è positivo`.

**Esercizio:** Modifica il codice precedente per gestire anche il caso in cui `numero` sia uguale a 0, usando `elif` (else if). Stampa "Il numero è zero" in quel caso.

In [ ]:
# Scrivi qui il tuo codice

**Riscontro atteso:** Se `numero = 0`, vedrai `Il numero è zero`. Se `numero = -5`, vedrai `Il numero non è positivo`.

### Ciclo `for`
Usato per iterare su una sequenza (come una lista, una stringa, o un range di numeri).

**Esercizio 1:** Usa un ciclo `for` e la funzione `range(5)` per stampare i numeri da 0 a 4.

In [ ]:
# Scrivi qui il tuo codice

**Riscontro atteso:** Dovresti vedere i numeri 0, 1, 2, 3, 4 stampati, ognuno su una nuova riga.

**Esercizio 2:** Data la stringa `testo = "Python"`, usa un ciclo `for` per stampare ogni carattere della stringa su una nuova riga.

In [ ]:
# Scrivi qui il tuo codice

**Riscontro atteso:** Dovresti vedere:
P
y
t
h
o
n

**Esercizio 3:** Data la lista `colori = ["rosso", "verde", "blu"]`, usa un ciclo `for` per stampare ogni colore della lista.

In [ ]:
# Scrivi qui il tuo codice

**Riscontro atteso:**
rosso
verde
blu

**Esercizio 4:** Usa un ciclo `for` e `range()` per stampare i numeri pari da 2 a 10 (inclusi).

In [ ]:
# Scrivi qui il tuo codice

**Riscontro atteso:**
2
4
6
8
10

**Esercizio 5:** Calcola la somma dei numeri da 1 a 5 usando un ciclo `for` e una variabile `somma_totale`. Stampa il risultato finale.

In [ ]:
# Scrivi qui il tuo codice

**Riscontro atteso:** Dovresti vedere `15`.

### Ciclo `while`
Esegue un blocco di codice fintanto che una condizione è vera.

**Esercizio:** Crea una variabile `contatore = 0`. Usa un ciclo `while` per stampare il valore di `contatore` e incrementarlo di 1, finché `contatore` è minore di 3.

In [ ]:
# Scrivi qui il tuo codice

**Riscontro atteso:** Dovresti vedere i numeri 0, 1, 2 stampati, ognuno su una nuova riga.

## 5. Liste

Le liste sono collezioni ordinate e modificabili di elementi. Gli elementi possono essere di tipi diversi. Si definiscono usando parentesi quadre `[]`.
Gli elementi di una lista sono accessibili tramite un **indice**, che parte da 0.

**Esercizio:**
1. Crea una lista chiamata `frutta` contenente "mela", "banana", "ciliegia".
2. Stampa la lista `frutta`.

In [14]:
# Scrivi qui il tuo codice
frutta = ["mela", "banana", "ciliegia"]
print(frutta)

['mela', 'banana', 'ciliegia']


**Riscontro atteso:** Dovresti vedere `['mela', 'banana', 'ciliegia']`.

**Esercizio:** Stampa il primo elemento della lista `frutta` (ricorda che l'indice parte da 0).

In [15]:
# Scrivi qui il tuo codice
print(frutta[0])

mela


**Riscontro atteso:** Dovresti vedere `mela`.

Puoi aggiungere elementi a una lista con il metodo `append()`.

**Esercizio:** Aggiungi "arancia" alla lista `frutta` e poi stampa la lista.

In [17]:
# Scrivi qui il tuo codice
frutta.append("arancia")
print(frutta)

['mela', 'banana', 'ciliegia', 'arancia', 'arancia']


**Riscontro atteso:** Dovresti vedere `['mela', 'banana', 'ciliegia', 'arancia']`.

### List Comprehension

La **list comprehension** è un modo conciso per creare liste in Python. La sintassi è:
```python
[espressione for elemento in sequenza if condizione]
```

Esempio pratico:
- Creare una lista dei quadrati dei numeri da 0 a 4:
```python
quadrati = [x**2 for x in range(5)]
# Risultato: [0, 1, 4, 9, 16]
```

**Esercizio 1:** Crea una lista chiamata `doppi` che contenga i doppi dei numeri da 1 a 5 utilizzando la list comprehension.
Stampa la lista ottenuta.


In [24]:
# Scrivi qui il tuo codice
doppi = [x*2 for x in range(1,6)]
print(doppi)

[2, 4, 6, 8, 10]


**Riscontro atteso:** Dovresti vedere [2, 4, 6, 8, 10]



**Esercizio 2:** Utilizza una list comprehension per creare una lista `solo_pari` che contenga solo i numeri pari da 0 a 10 inclusi. Stampa la lista.


In [31]:
# Scrivi qui il tuo codice
solo_pari = [x for x in range(11) if x % 2 == 0]
print(solo_pari)

[0, 2, 4, 6, 8, 10]


**Riscontro atteso:** Dovresti vedere [0, 2, 4, 6, 8, 10]



## 6. Dizionari

I dizionari sono collezioni non ordinate di coppie **chiave-valore**. Sono modificabili e non ammettono chiavi duplicate. Si definiscono usando parentesi graffe `{}`.

**Esercizio 1:**
1. Crea un dizionario chiamato `persona` con le seguenti chiavi e valori:
   - `"nome"`: `"Mario"`
   - `"eta"`: `30`
   - `"citta"`: `"Roma"`
2. Stampa il dizionario `persona`.

In [32]:
# Scrivi qui il tuo codice
persona = {"nome": "Mario",
           "età": 30,
           "città": "Roma"
           }

print(persona)

{'nome': 'Mario', 'età': 30, 'città': 'Roma'}


**Riscontro atteso:** Dovresti vedere qualcosa di simile a `{'nome': 'Mario', 'eta': 30, 'citta': 'Roma'}` (l'ordine potrebbe variare).

Puoi accedere ai valori usando la chiave corrispondente.

**Esercizio 2:** Stampa il valore associato alla chiave `"nome"` nel dizionario `persona`.

In [36]:
# Scrivi qui il tuo codice
print(persona["nome"])

Mario


**Riscontro atteso:** Dovresti vedere `Mario`.

Puoi aggiungere una nuova coppia chiave-valore o modificare un valore esistente.

**Esercizio 3:** Aggiungi al dizionario `persona` la chiave `"professione"` con valore `"Ingegnere"` e poi stampa il dizionario.

In [37]:
# Scrivi qui il tuo codice
persona["professione"] = "Ingegnere"

**Riscontro atteso:** Dovresti vedere `{'nome': 'Mario', 'eta': 30, 'citta': 'Roma', 'professione': 'Ingegnere'}`.

**Esercizio 4:** Itera e stampa tutte le **chiavi** del dizionario `persona`.

In [39]:
# Scrivi qui il tuo codice
for i in persona:
    print(i, persona[i])

nome Mario
età 30
città Roma
professione Ingegnere


**Riscontro atteso:** (l'ordine può variare)
nome
eta
citta
professione

**Esercizio 5:** Itera e stampa tutti i **valori** del dizionario `persona`.

In [40]:
# Scrivi qui il tuo codice
for i in persona.values():
    print(i, persona)

Mario {'nome': 'Mario', 'età': 30, 'città': 'Roma', 'professione': 'Ingegnere'}
30 {'nome': 'Mario', 'età': 30, 'città': 'Roma', 'professione': 'Ingegnere'}
Roma {'nome': 'Mario', 'età': 30, 'città': 'Roma', 'professione': 'Ingegnere'}
Ingegnere {'nome': 'Mario', 'età': 30, 'città': 'Roma', 'professione': 'Ingegnere'}


**Riscontro atteso:** (l'ordine può variare)
Mario
30
Roma
Ingegnere

**Esercizio 6:** Itera e stampa tutte le coppie **chiave-valore** del dizionario `persona` nel formato "chiave: valore".

In [ ]:
# Scrivi qui il tuo codice

**Riscontro atteso:** (l'ordine può variare)
nome: Mario
eta: 30
citta: Roma
professione: Ingegnere

**Esercizio 7:** Controlla se la chiave `"eta"` esiste nel dizionario `persona` e stampa un messaggio appropriato (`"L'età è presente"` o `"L'età non è presente"`).

In [ ]:
# Scrivi qui il tuo codice

**Riscontro atteso:** Dovresti vedere `L'età è presente`.

## 7. Funzioni

Le funzioni sono blocchi di codice riutilizzabili che eseguono un compito specifico. Si definiscono usando la parola chiave `def` seguita dal nome della funzione e parentesi `()`.
Possono accettare **argomenti** (o parametri) e possono restituire un valore usando `return`.

**Esercizio 1:** Definisci una funzione chiamata `saluta` che non accetta argomenti e stampa "Ciao a tutti!".

In [ ]:
# Scrivi qui il tuo codice

Per eseguire una funzione, devi **chiamarla**.

**Esercizio 2:** Chiama la funzione `saluta` che hai appena definito.

In [ ]:
# Scrivi qui il tuo codice

**Riscontro atteso:** Dovresti vedere `Ciao a tutti!`.

**Esercizio 3 (Focus su `return`):** 
1. Definisci una funzione chiamata `calcola_area_rettangolo` che accetta due argomenti: `base` e `altezza`.
2. All'interno della funzione, calcola l'area (`base * altezza`).
3. Usa la keyword `return` per restituire il valore dell'area calcolata.
4. Chiama la funzione con una base di 5 e un'altezza di 10.
5. Assegna il risultato restituito dalla funzione a una variabile chiamata `risultato_area`.
6. Stampa il valore di `risultato_area`.

In [ ]:
# Scrivi qui il tuo codice

**Riscontro atteso:** Dovresti vedere `50`.

**Esercizio 4:** Definisci una funzione chiamata `somma_due_numeri` che accetta due argomenti (`num1`, `num2`), **restituisce** la loro somma. Poi chiama la funzione con i valori 7 e 8, e stampa direttamente il risultato della chiamata.

In [ ]:
# Scrivi qui il tuo codice

**Riscontro atteso:** Dovresti vedere `15`.

## 8. Commenti

I commenti sono usati per spiegare il codice. Python ignora i commenti durante l'esecuzione.
- Un commento su una singola riga inizia con `#`.
- Un commento su più righe (o una docstring per funzioni/classi) può essere racchiuso tra triple virgolette `""" ... """` o `''' ... '''`.

**Esercizio:** Aggiungi un commento alla riga di codice sottostante che spieghi cosa fa.

In [ ]:
nome_utente = "Alice"
print(nome_utente)

**Riscontro atteso:** Il codice dovrebbe funzionare come prima, stampando `Alice`, ma ora hai un commento che non influisce sull'output.

## Complimenti!

Hai completato le basi di Python!

Buona programmazione!